### PR TITLE
Make category and transaction methods async

### PR DESCRIPTION
**Pull request by antonmironovv**

### CODE REVIEW

File: MoneySaver.Api.Services/Contracts/ITransactionCategoryService.cs [hunk 1]
```diff
@@ -9,9 +9,9 @@ namespace MoneySaver.Api.Services.Contracts
     public interface ITransactionCategoryService
     {
         Task<IEnumerable<TransactionCategoryModel>> GetAllCategoriesAsync();
-        TransactionCategoryModel GetCategory(int id);
-        TransactionCategoryModel UpdateCategory(TransactionCategoryModel categoryModel);
-        TransactionCategoryModel CreateCategory(TransactionCategoryModel categoryModel);
-        void RemoveCategory(int id);
+        Task <TransactionCategoryModel> GetCategoryAsync(int id);
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Contracts/ITransactionCategoryService.cs]: You can remove the space.

File: MoneySaver.Api.Services/Contracts/ITransactionCategoryService.cs [hunk 2]
```diff
@@ -9,9 +9,9 @@ namespace MoneySaver.Api.Services.Contracts
...
+        Task<TransactionCategoryModel> UpdateCategoryAsync(TransactionCategoryModel categoryModel);
+        Task<TransactionCategoryModel> CreateCategoryAsync(TransactionCategoryModel categoryModel);
+        void RemoveCategoryAsync(int id);
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Contracts/ITransactionCategoryService.cs]: Instead "void", it should be "Task" as this method is async

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs [hunk 3]
```diff
@@ -23,10 +23,18 @@ public TransactionCategoryService(IRepository<TransactionCategory> categoryRepos
...
-            this.categoryRepository.AddAsync(transactionCategory);
+            try
+            {
+                TransactionCategory transactionCategory = mapper.Map<TransactionCategory>(categoryModel);
+                await this.categoryRepository.AddAsync(transactionCategory);
+            }
+
+            catch (Exception ex)
+            {
+                ;
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs:40]: You need to return something different ( null for example) if an exception arises.

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs [hunk 4]
```diff
@@ -23,10 +23,18 @@ public TransactionCategoryService(IRepository<TransactionCategory> categoryRepos
...
 
-        public TransactionCategoryModel CreateCategory(TransactionCategoryModel categoryModel)
+        public async Task<TransactionCategoryModel> CreateCategoryAsync(TransactionCategoryModel categoryModel)
         {
-            TransactionCategory transactionCategory = mapper.Map<TransactionCategory>(categoryModel);
-            this.categoryRepository.AddAsync(transactionCategory);
+            try
+            {
+                TransactionCategory transactionCategory = mapper.Map<TransactionCategory>(categoryModel);
+                await this.categoryRepository.AddAsync(transactionCategory);
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs]: The id value from the newly created CategoryTransaction should be set in the TransactionCategoryModel too, before return.

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs [hunk 5]
```diff
@@ -73,26 +81,52 @@ public async Task<IEnumerable<TransactionCategoryModel>> GetAllCategoriesAsync()
...
+            catch (Exception ex)
+            {
+                ;
+            }
 
             return transactionCategoryModel;
         }
 
-        public void RemoveCategory(int id)
+        public async void RemoveCategoryAsync(int id)
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs]: The return type should be Task, not void as this is async method.

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs [hunk 6]
```diff
@@ -73,26 +81,52 @@ public async Task<IEnumerable<TransactionCategoryModel>> GetAllCategoriesAsync()
...
+
+            try
+            {
+                TransactionCategory transactionCategory = await this.categoryRepository.GetAll().FirstOrDefaultAsync(c => c.TransactionCategoryId == id);
+                transactionCategoryModel = mapper.Map<TransactionCategoryModel>(transactionCategory);
+            }
+
+            catch (Exception ex)
+            {
+                ;
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs]: As mentioned above about what to return when exception arises.

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs [hunk 7]
```diff
@@ -73,26 +81,52 @@ public async Task<IEnumerable<TransactionCategoryModel>> GetAllCategoriesAsync()
...
-            this.categoryRepository.UpdateAsync(transactionCategory);
+            try
+            {
+                TransactionCategory transactionCategory = mapper.Map<TransactionCategory>(categoryModel);
+                await this.categoryRepository.UpdateAsync(transactionCategory);
+            }
+
+            catch (Exception ex)
+            {
+                ;
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs:138]: As mentioned above about what to return when exception arises.

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs [hunk 8]
```diff
@@ -73,26 +81,52 @@ public async Task<IEnumerable<TransactionCategoryModel>> GetAllCategoriesAsync()
...
-            TransactionCategory transactionCategory = this.categoryRepository.GetAll().FirstOrDefault(c => c.TransactionCategoryId == id);
-            transactionCategory.IsDeleted = true;
-            transactionCategory.DeletedOnUtc = DateTime.UtcNow;
-            this.categoryRepository.RemoveAsync(transactionCategory);
+            try
+            {
+                TransactionCategory transactionCategory = await this.categoryRepository.GetAll().FirstOrDefaultAsync(c => c.TransactionCategoryId == id);
+                transactionCategory.IsDeleted = true;
+                transactionCategory.DeletedOnUtc = DateTime.UtcNow;
+                this.categoryRepository.RemoveAsync(transactionCategory);
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs]: "await" is missing.

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs [hunk 9]
```diff
@@ -73,26 +81,52 @@ public async Task<IEnumerable<TransactionCategoryModel>> GetAllCategoriesAsync()
...
+            }
+
+            catch (Exception ex)
+            {
+                ;
```
Comment (reb0ot1) [inline MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs:122]: As mentioned above about what to return when exception arises.

File: MoneySaver.Api/Controllers/CategoryController.cs [hunk 10]
```diff
@@ -38,27 +38,27 @@ IEnumerable<TransactionCategoryModel> result
         }
 
         [HttpGet("{transactionCategoryId}")]
-        public IActionResult GetCategory(int transactionCategoryId)
+        public async Task<IActionResult> GetCategoryAsync(int transactionCategoryId)
         {
-            return this.Ok(this.categoryService.GetCategory(transactionCategoryId));
+            return this.Ok(await this.categoryService.GetCategoryAsync(transactionCategoryId));
```
Comment (reb0ot1) [inline MoneySaver.Api/Controllers/CategoryController.cs]: You can put "await this.categoryService.GetCategoryAsync(transactionCategoryId)" on a separate row in order to be easier for debug and more readable.
for example:
var result = await this.categoryService.GetCategoryAsync(transactionCategoryId)
return this.Ok(result)

File: MoneySaver.Api/Controllers/CategoryController.cs [hunk 11]
```diff
@@ -38,27 +38,27 @@ IEnumerable<TransactionCategoryModel> result
...
-        public IActionResult CreateCategory(TransactionCategoryModel transactionCategoryModel)
+        public async Task<IActionResult> CreateCategoryAsync(TransactionCategoryModel transactionCategoryModel)
         {
-            return this.Ok(this.categoryService.CreateCategory(transactionCategoryModel));
+            return this.Ok(await this.categoryService.CreateCategoryAsync(transactionCategoryModel));
         }
 
         [HttpDelete]
-        public void RemoveTransaction(int transactionCategoryId)
+        public async void RemoveCategoryAsync(int transactionCategoryId)
```
Comment (reb0ot1) [inline MoneySaver.Api/Controllers/CategoryController.cs]: The return type should be Task not void.

File: MoneySaver.Api/Controllers/CategoryController.cs [hunk 12]
```diff
@@ -38,27 +38,27 @@ IEnumerable<TransactionCategoryModel> result
...
         {
-            this.categoryService.RemoveCategory(transactionCategoryId);
+             this.categoryService.RemoveCategoryAsync(transactionCategoryId);
```
Comment (reb0ot1) [inline MoneySaver.Api/Controllers/CategoryController.cs]: "await" is missing.

File: MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs (hunk 4 above)
Comment (antonmironovv) [reply MoneySaver.Api.Services/Implementation/TransactionCategoryService.cs]: The Mapper should set the TransactionCategoryId out of the box, or I am wrong? Feel free to correct me.

**Pull request merged**